beacon/src/beacon.rs

```diff
@@ -20,34 +21,53 @@ pub struct Beacon {
 }
 
 impl Beacon {
+    /// Construct a new beacon with a given remote and local entropy. The remote
+    /// and local entropy are checked for version equality.
+    ///
+    /// Version 0 beacons use a Sha256 of the remote and local entropy (data and
+    /// timestamp), which is then used as a 32 byte seed to a ChaCha12 rng. This
+    /// rng is used to choose a random frequency from the given region
+    /// parameters and a payload size between MIN_BEACON_V0_PAYLOAD_SIZE and
+    /// MAX_BEACON_V0_PAYLOAD_SIZE.
     pub fn new(
         remote_entropy: Entropy,
         local_entropy: Entropy,
         region_params: &[BlockchainRegionParamV1],
     ) -> Result<Self> {
-        let mut data = {
-            let mut hasher = Sha256::new();
-            remote_entropy.digest(&mut hasher);
-            local_entropy.digest(&mut hasher);
-            hasher.finalize().to_vec()
-        };
-        let mut seed = [0u8; 32];
-        seed.copy_from_slice(&data[0..32]);
-        let mut rng = rand_chacha::ChaCha12Rng::from_seed(seed);
+        if remote_entropy.version != local_entropy.version {
+            return Err(Error::invalid_version());
+        }
 
-        let frequency = rand_frequency(region_params, &mut rng)?;
-        let datarate = DataRate::Sf7bw125;
+        match remote_entropy.version {
+            0 => {
+                let mut data = {
+                    let mut hasher = Sha256::new();
+                    remote_entropy.digest(&mut hasher);
+                    local_entropy.digest(&mut hasher);
+                    hasher.finalize().to_vec()
+                };
+                let mut seed = [0u8; 32];
+                seed.copy_from_slice(&data[0..32]);
+                let mut rng = rand_chacha::ChaCha12Rng::from_seed(seed);
 
-        Ok(Self {
-            data: {
-                data.truncate(BEACON_PAYLOAD_SIZE);
-                data
-            },
-            frequency,
-            datarate,
-            local_entropy,
-            remote_entropy,
-        })
+                let frequency = rand_frequency(region_params, &mut rng)?;
+                let payload_size =
+                    rng.gen_range(MIN_BEACON_V0_PAYLOAD_SIZE..MAX_BEACON_V0_PAYLOAD_SIZE);
```

Comment:
Unless gen_range does something special, it'll never pick `MAX_BEACON_V0_PAYLOAD_SIZE` since ranges in rust are not inclusive.

Reply:
you're right.. fixing